README.md
# py-ard

Swiss army knife of **HLA** Nomenclature

[![PyPi Version](https://img.shields.io/pypi/v/py-ard.svg)](https://pypi.python.org/pypi/py-ard)

![py-ard-logo.png](images/py-ard-logo.png)

### `py-ard` is ARD reduction for HLA in Python

Human leukocyte antigen (HLA) genes encode cell surface proteins that are important for immune regulation. Exons encoding the Antigen Recognition Domain (ARD) are the most polymorphic region of HLA genes and are important for donor/recipient [HLA matching](https://bethematch.org/patients-and-families/before-transplant/find-a-donor/hla-matching/). The history of allele typing methods has played a major role in determining resolution and ambiguity of reported HLA values. Although HLA [nomenclature](https://www.theatlantic.com/magazine/archive/2023/04/clint-smith-nomenclature-poem/673097/) has not always conformed to the same standard, it is now defined by [The WHO Nomenclature Committee for Factors of the HLA System](https://hla.alleles.org/nomenclature/committee.html). `py-ard` is aware of the variation in historical resolutions and grouping and is able to translate from one representation to another based on alleles published quarterly by [IPD/IMGT-HLA](https://github.com/ANHIG/IMGTHLA/).


## Table of Contents
1. [Installation](#installation)
    * [Install From PyPi](#install-from-pypi)
    * [Install With Homebrew](#install-with-homebrew)
    * [Install From Source](#install-from-source)
2. [Using `py-ard`](#using-py-ard)
    * [Using `py-ard` from Python](#using-py-ard-from-python-code)
    * [Using `py-ard` from R](#using-py-ard-from-r-code)
    * [Perform Reduxtion](#reduce-typings)
    * [DRBX blending](#perform-drb1-blending-with-drb3-drb4-and-drb5)
    * [Expand/Lookup MAC](#mac-codes)
3. [Command Line Tools](#command-line-tools)
    * [`pyard-import` Import Reference Data](#pyard-import-import-the-latest-ipd-imgthla-database)
    * [`pyard-status` Show Statuses of Databases](#pyard-status-show-database-status)
    * [`pyard` Redux](#pyard-redux-quickly)
    * [`pyard-csv-reduce` Batch Mode Redux](#pyard-csv-reduce-batch-reduce-a-csv-file)
4. [`py-ard` REST Webservice](#py-ard-rest-web-service)
5. [Docker Deployment](#docker-deployment-of-py-ard-rest-web-service)

## Installation
`py-ard` works with Python 3.8 and higher.

### Install from PyPi

```shell
pip install py-ard
```
Note: With `py-ard` version *1.0.0* and higher, the redux API has changed. If your use requires the older API, please install with `pip install py-ard==0.9.2`


### Install With Homebrew

On macOS, `py-ard` can be installed using Homebrew package manager.
This is very handy for using the command line versions of the tool without having to create virtual environments.

First time, you'd need to tap the `nmdp-bioinformatics` tap.

```shell
brew tap nmdp-bioinformatics/tap
```

Install `py-ard`

```shell
brew install py-ard
```

Homebrew will notify you as new versions of `py-ard` are released.

### Install from source

```shell
python3 -m venv venv
source venv/bin/activate

python setup.py install
```

See [Our Contribution Guide](CONTRIBUTING.rst) for open source contribution to `py-ard`.

## Using `py-ard`

### Using `py-ard` from Python code

`py-ard` can be used in a program to reduce/expand HLA GL String representation. If pyard discovers an invalid Allele, it'll throw an Invalid Exception, not silently return an empty result.

#### Initialize `py-ard`

Import `pyard` package.

```python
import pyard
```


Initialize `ARD` object with a version of IPD-IMGT/HLA database

```python
import pyard

ard = pyard.init('3510')
```

When processing a large numbers of typings, it's helpful to have a cache of previously calculated reductions to make similar typings reduce faster. The cache size of pre-computed reductions can be changed from the default of 1,000 by setting `cache_size` argument. This increases the memory footprint but will significantly increase the processing times for large number of reductions.

```python
import pyard

max_cache_size = 1_000_000
ard = pyard.init('3510', cache_size=max_cache_size)
```

By default, the IPD-IMGT/HLA data is stored locally in `$TMPDIR/pyard`. This may be removed when your computer restarts. You can specify a different, more permanent directory for the cached data.

```python
import pyard.ard

ard = pyard.init('3510', data_dir='/tmp/py-ard')
```

As MAC data changes frequently, you can choose to refresh the MAC code for current IPD-IMGT/HLA database version.

```python
ard.refresh_mac_codes()
```

The default initialization is to use the latest version of IPD-IMGT/HLA database.

```python
import pyard

ard = pyard.init()
```

You can check the current version of IPD-IMGT/HLA database.
```python
ard.get_db_version()
```

### Reduce Typings

**Note**: Previous to version of 1.0.0 release of `py-ard`, there was `redux` and `redux_gl` methods on `ard`. They have been consolidated so that `redux` handles both GL Strings and individual alleles.

Reduce a single locus HLA Typing by specifying the allele/MAC/XX code and the reduction method to `redux` method.

```python
allele = "A*01:01:01"

ard.redux(allele, 'G')
# >>> 'A*01:01:01G'

ard.redux(allele, 'lg')
# >>> 'A*01:01g'

ard.redux(allele, 'lgx')
# >>> 'A*01:01'
```

Reduce an ambiguous GL String

```python
# Reduce GL String
#
ard.redux("A*01:01/A*01:01N+A*02:AB^B*07:02+B*07:AB", "G")
# 'B*07:02:01G+B*07:02:01G^A*01:01:01G+A*02:01:01G/A*02:02'
```

You can also reduce serology based typings.

```python
ard.redux('B14', 'lg')
# >>> 'B*14:01g/B*14:02g/B*14:03g/B*14:04g/B*14:05g/B*14:06g/B*14:08g/B*14:09g/B*14:10g/B*14:11g/B*14:12g/B*14:13g/B*14:14g/B*14:15g/B*14:16g/B*14:17g/B*14:18g/B*14:19g/B*14:20g/B*14:21g/B*14:22g/B*14:23g/B*14:24g/B*14:25g/B*14:26g/B*14:27g/B*14:28g/B*14:29g/B*14:30g/B*14:31g/B*14:32g/B*14:33g/B*14:34g/B*14:35g/B*14:36g/B*14:37g/B*14:38g/B*14:39g/B*14:40g/B*14:42g/B*14:43g/B*14:44g/B*14:45g/B*14:46g/B*14:47g/B*14:48g/B*14:49g/B*14:50g/B*14:51g/B*14:52g/B*14:53g/B*14:54g/B*14:55g/B*14:56g/B*14:57g/B*14:58g/B*14:59g/B*14:60g/B*14:62g/B*14:63g/B*14:65g/B*14:66g/B*14:68g/B*14:70Qg/B*14:71g/B*14:73g/B*14:74g/B*14:75g/B*14:77g/B*14:82g/B*14:83g/B*14:86g/B*14:87g/B*14:88g/B*14:90g/B*14:93g/B*14:94g/B*14:95g/B*14:96g/B*14:97g/B*14:99g/B*14:102g'
```

## Valid Reduction Types

| Reduction Type | Description                                     |
|----------------|-------------------------------------------------|
| `G`            | Reduce to G Group Level                         |
| `P`            | Reduce to P Group Level                         |
| `lg`           | Reduce to 2 field ARD level (append `g`)        |
| `lgx`          | Reduce to 2 field ARD level                     |
| `W`            | Reduce/Expand to 3 field WHO nomenclature level |
| `exon`         | Reduce/Expand to exon level                     |
| `U2`           | Reduce to 2 field unambiguous level             |
| `S`            | Reduce to Serological level                     |

### Perform DRB1 blending with DRB3, DRB4 and DRB5

```python
import pyard

pyard.dr_blender(drb1='HLA-DRB1*03:01+DRB1*04:01', drb3='DRB3*01:01', drb4='DRB4*01:03')
# >>> 'DRB3*01:01+DRB4*01:03'
```

## MAC Codes

`py-ard` supports not only reducing to various types but helps in expanding and
looking up MAC representation. See [MAC Service UI](https://hml.nmdp.org/MacUI/) for detail.

### Expand MAC

You can also use `py-ard` to expand MAC codes. Use `expand_mac` method on `ard`.
```python
ard.expand_mac('HLA-A*01:BC')
# 'HLA-A*01:02/HLA-A*01:03'
```

### Lookup MAC

Find the corresponding MAC code for an allele list GL String.

```python
ard.lookup_mac('A*01:02/A*01:01/A*01:03')
# A*01:MN
```

### CWD Reduction

Reduce a MAC code or an allele list GL String to CWD reduced list.
```python
ard.cwd_redux("B*15:01:01/B*15:01:03/B*15:04/B*15:07/B*15:26N/B*15:27")
# => B*15:01/B*15:07
```

The above 2 methods can be chained to get back a MAC code that has a CWD reduced version.

```python
ard.lookup_mac(ard.cwd_redux("B*15:01:01/B*15:01:03/B*15:04/B*15:07/B*15:26N/B*15:27"))
# 'B*15:AH'
```

### Using `py-ard` from R code

`py-ard` works well from `R` as well. Please see [Using pyard from R language](https://github.com/nmdp-bioinformatics/py-ard/wiki/Using-pyard-library-from-R-language) page for detailed walkthrough.

## Command Line Tools

Various command line interface (CLI) tools are available to use for managing local IPD-IMGT/HLA cache database, running impromptu reduction queries and batch processing of CSV files.

For all tools, use `--ipd-version` and `--data-dir` to specify the IPD-IMGT/HLA database version and the directory where the SQLite files are created.

### `pyard-import` Import the latest IPD-IMGT/HLA database

`pyard-import` helps with importing and reinstalling of prepared IPD-IMGT/HLA and MAC data.

Use `pyard-import -h` to see all the options available.
```shell
$ pyard-import -h
usage: pyard-import [-h] [--list] [-i IMGT_VERSION] [-d DATA_DIR] [--v2-to-v3-mapping V2_V3_MAPPING] [--refresh-mac] [--re-install] [--skip-mac]


CHANGELOG.md
<a name="1.0.6"></a>
# [1.0.6 Validation of allele specific MAC codes](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.6) - 30 Oct 2023

- Use allele specific antigen code rules when validating MACs that cross antigen group similar to [MAC Service](https://hml.nmdp.org/macui/)
- Returns the original `InvalidAlleleError` instead of wrapping it in `InvalidTypingError` when an allele is not valid.

<a name="1.0.5"></a>
# [1.0.5 Non strict mode](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.5) - 04 Oct 2023

Supports non-strict mode makes valid alleles by adding expression characters to invalid alleles.

## Use non `strict` mode in config to reduce alleles that may be valid with expression characters.

```python
>>> my_configs = {'strict': False, 'verbose_log': True}
>>> import pyard
>>> ard = pyard.init(config=my_configs, load_mac=False)

>>> ard.redux('A*24:329', 'lgx')
A*24:329 is not valid. Using A*24:329Q
'A*24:329Q'

>>> ard.redux('DQB1*03:276', 'lgx')
DQB1*03:276 is not valid. Using DQB1*03:276N
'DQB1*03:01'
```

## Add non-strict and verbose modes to pyard CLI.

```bash
❯ pyard --gl "DQB1*03:276" -r lgx
Typing Error: DQB1*03:276 is not valid GL String.
 DQB1*03:276 is not a valid Allele

❯ pyard --non-strict --gl "DQB1*03:276" -r lgx
DQB1*03:01

❯ pyard --non-strict --verbose --gl "DQB1*03:276" -r lgx
DQB1*03:276 is not valid. Using DQB1*03:276N
DQB1*03:01
```


[Changes][1.0.5]


<a name="1.0.4"></a>
# [Fixes when used without login user (1.0.4)](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.4) - 19 Sep 2023

Fixed the OSError when running without a login user.

[Changes][1.0.4]


<a name="1.0.3"></a>
# [1.0.3 – Permission Errors and pyard updates](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.3) - 17 Aug 2023

- Alert permission Errors when `data_dir` is not writable
- Add `--lookup-mac` and `--expand-mac` to `pyard` command

[Changes][1.0.3]


<a name="1.0.2"></a>
# [Fixes issue with using py-ard without MAC (1.0.2)](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.2) - 03 Aug 2023

Fixes issue with using py-ard without MAC
 - When using py-ard with `load_mac=False`, check if the allele looks like a MAC

[Changes][1.0.2]


<a name="1.0.1"></a>
# [1.0.1 → Bug Fixes for 1.0.0 Release](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.1) - 23 Jun 2023

Fixed bugs and added some niceties

[#237](https://github.com/nmdp-bioinformatics/py-ard/issues/237)	Additional functionalities for `pyard` CLI command	enhancement
[#235](https://github.com/nmdp-bioinformatics/py-ard/issues/235)	Invalid allele in CWD2 mapping	bug
[#234](https://github.com/nmdp-bioinformatics/py-ard/issues/234)	Success with lgx when given Invalid format	bug
[#233](https://github.com/nmdp-bioinformatics/py-ard/issues/233)	`validate` endpoint can be a GET request	enhancement
[#230](https://github.com/nmdp-bioinformatics/py-ard/issues/230)	Redux with a "P" option does not return P-groups for two-field or three-field alleles	bug


[Changes][1.0.1]


<a name="1.0.0"></a>
# [Final Release 1.0.0](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.0) - 04 May 2023

Release YAY!! 🎉

[Changes][1.0.0]


<a name="1.0.0rc7"></a>
# [Fix PyPi Packaging (1.0.0rc7)](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.0rc7) - 17 Apr 2023

Include `requirements.txt` files

[Changes][1.0.0rc7]


<a name="1.0.0rc6"></a>
# [Batch Reduce Fixes (1.0.0rc6)](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.0rc6) - 05 Apr 2023

Bug fixes
- Fix issue when adding locus name to a serology value
- Fix issue when reducing an allele-list
- Account for lower case allele names

[Changes][1.0.0rc6]


<a name="1.0.0rc5"></a>
# [Decode MAC and CWD 2.0 Reductions (1.0.0rc5)](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.0rc5) - 03 Apr 2023

Support looking up MAC code from allele list with `lookup_mac` method.

Support CWD Reduction:
 - add `cwd_redux` method to produce CWD only allele list
 - create `cwd2` table on load
 - `cwd-redux` endpoint to allow CWD reduction

[Changes][1.0.0rc5]


<a name="0.9.2"></a>
# [Fix Pandas dependencies Latest (0.9.2)](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/0.9.2) - 23 Mar 2023

Set to `pandas==1.5.3`

[Changes][0.9.2]


<a name="1.0.0rc4"></a>
# [Documentation Update and Cleanup (1.0.0rc4)](https://github.com/nmdp-bioinformatics/py-ard/releases/tag/1.0.0rc4) - 15 Mar 2023

## What's Changed

pyard/db.py
# -*- coding: utf-8 -*-
#
#    py-ard
#    Copyright (c) 2023 Be The Match operated by National Marrow Donor Program. All Rights Reserved.
#
#    This library is free software; you can redistribute it and/or modify it
#    under the terms of the GNU Lesser General Public License as published
#    by the Free Software Foundation; either version 3 of the License, or (at
#    your option) any later version.
#
#    This library is distributed in the hope that it will be useful, but WITHOUT
#    ANY WARRANTY; with out even the implied warranty of MERCHANTABILITY or
#    FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#    License for more details.
#
#    You should have received a copy of the GNU Lesser General Public License
#    along with this library;  if not, write to the Free Software Foundation,
#    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA.
#
#    > http://www.fsf.org/licensing/licenses/lgpl.html
#    > http://www.opensource.org/licenses/lgpl-license.php
#
import os
import pathlib
import sqlite3
import sys
from typing import Tuple, Dict, Set, List

from .mappings import ARSMapping, CodeMappings, AlleleGroups
from .misc import get_ipd_db_versions, get_default_db_directory


def create_db_connection(data_dir, ipd_version, ro=False):
    """
    Create a  connection to a sqlite database in read-only mode
    or read-write mode (default)

    :param data_dir: The directory where the db is/will be created
    :param ipd_version: IPD-IMGT/HLA db version
    :param ro: Read-only mode ?
    :return: db connection of type sqlite.Connection
    """
    # Set data directory where all the downloaded files will go
    if data_dir is None:
        data_dir = get_default_db_directory()

    db_filename = f"{data_dir}/pyard-{ipd_version}.sqlite3"

    if ro:
        # If in read-only mode, make sure the db file exists
        if not pathlib.Path(db_filename).exists():
            raise RuntimeError(f"Reference Database {db_filename}  not available.")
        # Open the database in read-only mode
        file_uri = f"file:{db_filename}?mode=ro"
        # Multiple threads can access the same connection since it's only ro
        return sqlite3.connect(file_uri, check_same_thread=False, uri=True), db_filename

    # Check the ipd_version is a valid IPD-IMGT/HLA DB Version
    # by querying the IPD-IMGT/HLA site
    if ipd_version != "Latest":
        if not pathlib.Path(db_filename).exists():
            all_ipd_versions = get_ipd_db_versions()
            if str(ipd_version) not in all_ipd_versions:
                raise ValueError(
                    f"{ipd_version} is not a valid IPD-IMGT/HLA database version."
                )

    # Create the data directory if it doesn't exist
    if not pathlib.Path(data_dir).exists():
        pathlib.Path(data_dir).mkdir(parents=True, exist_ok=True)

    # Check for permission to read/write
    if pathlib.Path(db_filename).exists():
        # Check file is readable
        if not os.access(db_filename, os.R_OK):
            print(
                f"Permission Error reading {db_filename}"
                "Please specify accessible --data-dir."
            )
            sys.exit(1)
    else:
        # Check directory is writeable
        if os.access(data_dir, os.W_OK):
            print(f"Creating {db_filename} as cache.")
        else:
            print(
                f"Directory {data_dir} is not writeable."
                "Please specify accessible --data-dir."
            )
            sys.exit(1)

    # Open the database for read/write
    file_uri = f"file:{db_filename}"
    return sqlite3.connect(file_uri, uri=True), db_filename


def table_exists(connection: sqlite3.Connection, table_name: str) -> bool:
    """
    Does the table exist in the database ?

    :param connection: db connection of type sqlite.Connection
    :param table_name: table in the sqlite db
    :return: bool indicating whether table_name exists as a table
    """
    query = """SELECT count(*) from sqlite_master where type = 'table' and name = ?"""
    cursor = connection.execute(query, (table_name,))
    result = cursor.fetchone()
    cursor.close()
    return result[0] > 0


def tables_exist(connection: sqlite3.Connection, table_names: List[str]):
    """
    Do all the given tables exist in the database ?

    :param connection: db connection of type sqlite.Connection
    :param table_names: names of tables in the sqlite db
    :return: bool indicating whether all table_names exists
    """
    return all([table_exists(connection, table_name) for table_name in table_names])


def count_rows(connection: sqlite3.Connection, table_name: str) -> int:
    """
    Count number of rows in the table.

    :param connection: db connection of type sqlite.Connection
    :param table_name: table in the sqlite db
    :return: bool indicating whether table_name exists as a table
    """
    query = f"SELECT count(*) from '{table_name}'"
    cursor = connection.execute(query)
    result = cursor.fetchone()
    cursor.close()
    return result[0]


def mac_code_to_alleles(connection: sqlite3.Connection, code: str) -> List[str]:
    """
    Look up the MAC code in the database and return corresponding list
    of alleles.
    :param connection: db connection of type sqlite.Connection
    :param code: MAC code
    :return: List of alleles
    """
    mac_query = "SELECT alleles from mac_codes where code = ?"
    cursor = connection.execute(mac_query, (code,))
    result = cursor.fetchone()
    cursor.close()
    if result:
        alleles = result[0].split("/")
    else:
        alleles = []
    return alleles


def alleles_to_mac_code(
    connection: sqlite3.Connection, code_expansion: str
) -> List[str]:
    """
    Look up the MAC code in the database and based on list of allele expansion
    :param connection: db connection of type sqlite.Connection
    :param code_expansion: expansion of MAC code
    :return: List of alleles
    """
    mac_query = "SELECT code from mac_codes where alleles = ?"
    cursor = connection.execute(mac_query, (code_expansion,))
    result = cursor.fetchone()
    cursor.close()
    if result:
        return result[0]
    return None


def serology_to_alleles(connection: sqlite3.Connection, serology: str) -> List[str]:
    """
    Look up Serology in the database and return corresponding list of alleles.

    :param connection: db connection of type sqlite.Connection
    :param serology: Serology
    :return: List of alleles
    """
    serology_query = "SELECT allele_list from serology_mapping where serology = ?"
    cursor = connection.execute(serology_query, (serology,))
    result = cursor.fetchone()
    cursor.close()
    if result:
        alleles = result[0].split("/")
    else:
        alleles = []
    return alleles


def is_valid_serology(connection: sqlite3.Connection, serology: str) -> bool:
    """
    Check db if the serology exists

    :param connection: db connection of type sqlite.Connection
    :param serology: serology to test
    :return: is it serology ?
    """
    serology_query = (
        "SELECT count(allele_list) from serology_mapping where serology = ?"
    )
    cursor = connection.execute(serology_query, (serology,))
    result = cursor.fetchone()
    cursor.close()
    return result[0] > 0


def v2_to_v3_allele(connection: sqlite3.Connection, v2_allele: str) -> str:
    """
    Look up V3 version of the allele in the database.

    :param connection: db connection of type sqlite.Connection
    :param v2_allele: V2 allele
    :return: V3 allele
    """
    v2_query = "SELECT v3 from v2_mapping where v2 = ?"
    cursor = connection.execute(v2_query, (v2_allele,))
    result = cursor.fetchone()
    cursor.close()
    if result:
        return result[0]
    return None


def save_dict(
    connection: sqlite3.Connection,
    table_name: str,
    dictionary: Dict[str, str],
    columns: Tuple[str, str],
) -> bool:
    """
    Save the dictionary as a table with column names from columns Tuple.

    :param connection: db connection of type sqlite.Connection
    :param table_name: name of the table to create
    :param dictionary: the dictionary which to take the key and values from
    :param columns: column names in the table
    :return: success status
    """
    cursor = connection.cursor()

    # Drop the table first
    drop_table_sql = f"DROP TABLE IF EXISTS {table_name}"
    cursor.execute(drop_table_sql)

    # Create table
    create_table_sql = f"""CREATE TABLE {table_name} (
                            {columns[0]} TEXT PRIMARY KEY,
                            {columns[1]} TEXT NOT NULL
                    )"""
    cursor.execute(create_table_sql)

    # insert
    cursor.executemany(f"INSERT INTO {table_name} VALUES (?, ?)", dictionary.items())

    # commit transaction - writes to the db
    connection.commit()
    # close the cursor
    cursor.close()

    return True


def save_set(
    connection: sqlite3.Connection, table_name: str, rows: Set, column: str
) -> bool:
    """
    Save the set rows to the table table_name in the column

    :param connection: db connection of type sqlite.Connection
    :param table_name: name of the table to create
    :param rows: set which will become the the column in the table
    :param column: name of the column in the table
    :return: success status
    """
    cursor = connection.cursor()

    # Drop the table first
    drop_table_sql = f"DROP TABLE IF EXISTS {table_name}"
    cursor.execute(drop_table_sql)

    # Create table
    create_table_sql = f"""CREATE TABLE {table_name} (
                            {column} TEXT PRIMARY KEY
                    )"""
    cursor.execute(create_table_sql)

    # insert
    cursor.executemany(
        f"INSERT INTO {table_name} VALUES (?)",
        zip(
            rows,
        ),
    )

    # commit transaction - writes to the db
    connection.commit()
    # close the cursor
    cursor.close()

    return True


def load_set(connection: sqlite3.Connection, table_name: str, column: str) -> Set:
    """
    Retrieve the first column of the table as a set

    :param connection: db connection of type sqlite.Connection
    :param table_name: name of the table to query
    :param column: name of the column in the table to query
    :return: set containing values from the column
    """
    cursor = connection.cursor()
    select_all_query = f"SELECT {column} FROM {table_name}"
    cursor.execute(select_all_query)
    table_as_set = set(map(lambda t: t[0], cursor.fetchall()))
    cursor.close()
    return table_as_set


def load_cwd(connection: sqlite3.Connection, locus: str) -> Set:
    """
    Retrieve the CWD Version 2 alleles for a locus as a set

    :param connection: db connection of type sqlite.Connection
    :param locus: name of the column in the table to query
    :return: CWD allele set
    """
    cursor = connection.cursor()
    select_all_query = f"SELECT allele FROM cwd2 where locus='{locus}'"
    cursor.execute(select_all_query)
    table_as_set = set(map(lambda t: t[0], cursor.fetchall()))
    cursor.close()
    return table_as_set


def load_dict(
    connection: sqlite3.Connection, table_name: str, columns: Tuple[str, str]
) -> Dict[str, str]:
    """
    Retrieve the values in columns as a name, value pair and create a dict.

    :param connection: db connection of type sqlite.Connection
    :param table_name: name of the table to query
    :param columns: column names in the table
    :return: a dict containing key,value pairs from the columns
    """
    cursor = connection.cursor()
    select_all_query = f"SELECT {columns[0]}, {columns[1]} FROM {table_name}"
    cursor.execute(select_all_query)
    table_as_dict = {k: v for k, v in cursor.fetchall()}
    cursor.close()
    return table_as_dict


def similar_alleles(connection: sqlite3.Connection, allele_prefix: str) -> Set[str]:
    """
    Find similar alleles starting with the provided prefix.

    :param connection: db connection of type sqlite.Connection
    :param allele_prefix: Allele name to use as a prefix to find similar alleles
    :return: list of similar alleles
    """
    query = "SELECT allele FROM alleles WHERE allele LIKE ?"
    cursor = connection.execute(query, (f"{allele_prefix}%",))
    result = cursor.fetchall()
    # fetchall() returns a list of tuples of results
    # e.g. [('C*04:09N',)]
    # Get out the first value of the tuple from the result list
    alleles = set(map(lambda t: t[0], result))
    return alleles


def similar_mac(connection: sqlite3.Connection, mac_prefix: str) -> Set[str]:
    """
    Find similar MAC codes starting with the provided prefix.

    :param connection: db connection of type sqlite.Connection
    :param mac_prefix: MAC fragment to use as a prefix to find similar MACs
    :return: list of similar MAC codes
    """
    query = "SELECT code FROM mac_codes WHERE code LIKE ?"
    cursor = connection.execute(query, (f"{mac_prefix}%",))
    result = cursor.fetchall()
    # fetchall() returns a list of tuples of results
    # e.g. [('DJZUP',)]
    # Get out the first value of the tuple from the result list
    codes = set(map(lambda t: t[0], result))
    return codes


def find_serology_for_allele(
    connection: sqlite3.Connection, allele_name: str
) -> Dict[str, str]:
    """
    Find similar alleles starting with the provided allele_name.

    :param connection: db connection of type sqlite.Connection
    :param allele_name: Allele name to use as a prefix to find similar alleles
    :return: list of similar alleles
    """
    query = (
        "SELECT serology, allele_list FROM serology_mapping WHERE allele_list LIKE ?"
    )
    cursor = connection.execute(query, (f"%{allele_name}%",))
    results = cursor.fetchall()
    # fetchall() returns a list of tuples of results
    # e.g. [('A1', ''A*01:01:01:01/A*01:01:01:03')]
    serology_mapping = {serology: allele_list for serology, allele_list in results}
    return serology_mapping


def get_user_version(connection: sqlite3.Connection) -> int:
    """
    Retrieve user_version from db

    :connection: sqlite3.Connection: SQLite DB Connection
    """
    query = "PRAGMA user_version"
    cursor = connection.execute(query)
    result = cursor.fetchone()
    version = result[0]
    cursor.close()

    if version:
        return version
    return None


def set_user_version(connection: sqlite3.Connection, version: int):
    """
    Save the version number as user_version in db

    :connection: sqlite3.Connection:
    :version: int: version number to store
    @return:
    """
    query = f"PRAGMA user_version={version}"
    cursor = connection.execute(query)
    # commit transaction - writes to the db
    connection.commit()
    # close the cursor
    cursor.close()


def load_ars_mappings(db_connection):
    dup_g = load_dict(db_connection, table_name="dup_g", columns=("allele", "g_group"))
    dup_lgx = load_dict(
        db_connection, table_name="dup_lgx", columns=("allele", "lgx_group")
    )
    g_group = load_dict(db_connection, table_name="g_group", columns=("allele", "g"))
    p_group = load_dict(db_connection, table_name="p_group", columns=("allele", "p"))
    lgx_group = load_dict(
        db_connection, table_name="lgx_group", columns=("allele", "lgx")
    )
    exon_group = load_dict(
        db_connection, table_name="exon_group", columns=("allele", "exon")
    )
    p_not_g = load_dict(db_connection, table_name="p_not_g", columns=("allele", "lgx"))
    return ARSMapping(
        dup_g=dup_g,
        dup_lgx=dup_lgx,
        g_group=g_group,
        p_group=p_group,
        lgx_group=lgx_group,
        exon_group=exon_group,
        p_not_g=p_not_g,
    )


def save_ars_mappings(db_connection: sqlite3.Connection, ars_mapping: ARSMapping):
    save_dict(
        db_connection,
        table_name="p_not_g",
        dictionary=ars_mapping.p_not_g,
        columns=("allele", "lgx"),
    )
    save_dict(
        db_connection,
        table_name="dup_g",
        dictionary=ars_mapping.dup_g,
        columns=("allele", "g_group"),
    )
    save_dict(
        db_connection,
        table_name="dup_lgx",
        dictionary=ars_mapping.dup_lgx,
        columns=("allele", "lgx_group"),
    )
    save_dict(
        db_connection,
        table_name="g_group",
        dictionary=ars_mapping.g_group,
        columns=("allele", "g"),
    )
    save_dict(
        db_connection,
        table_name="p_group",
        dictionary=ars_mapping.p_group,
        columns=("allele", "p"),
    )
    save_dict(
        db_connection,
        table_name="lgx_group",
        dictionary=ars_mapping.lgx_group,
        columns=("allele", "lgx"),
    )
    save_dict(
        db_connection,
        table_name="exon_group",
        dictionary=ars_mapping.exon_group,
        columns=("allele", "exon"),
    )


def save_code_mappings(
    db_connection,
    exp_alleles,
    flat_who_group,
    flat_xx_codes,
    valid_alleles,
    who_alleles,
):
    save_dict(db_connection, "exp_alleles", exp_alleles, ("exp_allele", "allele_list"))
    save_dict(db_connection, "xx_codes", flat_xx_codes, ("allele_1d", "allele_list"))
    # Save this version of the valid alleles
    save_set(db_connection, "alleles", valid_alleles, "allele")
    # Save this version of the WHO alleles
    save_set(db_connection, "who_alleles", who_alleles, "allele")
    save_dict(
        db_connection, "who_group", flat_who_group, columns=("who", "allele_list")
    )


def load_code_mappings(db_connection):
    valid_alleles = load_set(db_connection, "alleles", "allele")
    who_alleles = load_set(db_connection, "who_alleles", "allele")
    who_group = load_dict(db_connection, "who_group", ("who", "allele_list"))
    who_group = {k: v.split("/") for k, v in who_group.items()}
    xx_codes = load_dict(db_connection, "xx_codes", ("allele_1d", "allele_list"))
    xx_codes = {k: v.split("/") for k, v in xx_codes.items()}
    exp_alleles = load_dict(db_connection, "exp_alleles", ("exp_allele", "allele_list"))
    exp_alleles = {k: v.split("/") for k, v in exp_alleles.items()}
    return (
        CodeMappings(xx_codes=xx_codes, who_group=who_group),
        AlleleGroups(
            alleles=valid_alleles, who_alleles=who_alleles, exp_alleles=exp_alleles
        ),
    )


def load_shortnulls(db_connection):
    shortnulls = load_dict(db_connection, "shortnulls", ("shortnull", "allele_list"))
    shortnulls = {k: v.split("/") for k, v in shortnulls.items()}
    return shortnulls


def save_shortnulls(db_connection, shortnulls):
    save_dict(db_connection, "shortnulls", shortnulls, ("shortnull", "allele_list"))


def save_mac_codes(db_connection, mac, mac_table_name):
    # Save the mac dict to db
    save_dict(
        db_connection,
        table_name=mac_table_name,
        dictionary=mac,
        columns=("code", "alleles"),
    )


def save_serology_mappings(db_connection, sero_mapping):
    # Save the serology mapping to db
    save_dict(
        db_connection,
        table_name="serology_mapping",
        dictionary=sero_mapping,
        columns=("serology", "allele_list"),
    )


def load_v2_v3_mappings(db_connection):
    # TODO: Create mapping table using both the allele list history and
    #  deleted alleles as reference.
    # Temporary Example
    v2_to_v3_example = {
        "A*0104": "A*01:04:01:01N",
        "A*0105N": "A*01:04:01:01N",
        "A*0111": "A*01:11N",
        "A*01123": "A*01:123N",
        "A*0115": "A*01:15N",
        "A*0116": "A*01:16N",
        "A*01160": "A*01:160N",
        "A*01162": "A*01:162N",
        "A*01178": "A*01:178N",
        "A*01179": "A*01:179N",
        "DRB5*02ZB": "DRB5*02:UTV",
    }
    save_dict(
        db_connection,
        table_name="v2_mapping",
        dictionary=v2_to_v3_example,
        columns=("v2", "v3"),
    )


def load_serology_broad_split_mappings(db_connection):
    sero_mapping = load_dict(
        db_connection, "serology_broad_split_mapping", ("serology", "splits")
    )
    sero_splits = {k: v.split("/") for k, v in sero_mapping.items()}
    return sero_splits


def save_serology_broad_split_mappings(db_connection, sero_mapping):
    # Save the `splits` as a "/" delimited string to db
    sero_splits = {sero: "/".join(splits) for sero, splits in sero_mapping.items()}
    save_dict(
        db_connection,
        table_name="serology_broad_split_mapping",
        dictionary=sero_splits,
        columns=("serology", "splits"),
    )


def save_cwd2(db_connection: sqlite3.Connection, cwd2_map: dict):
    save_dict(
        db_connection,
        table_name="cwd2",
        dictionary=cwd2_map,
        columns=("allele", "locus"),
    )
